runtime/sema/check_variable_declaration.go

```diff
@@ -234,3 +236,119 @@ func (checker *Checker) elaborateNestedResourceMoveExpression(expression ast.Exp
 		checker.Elaboration.IsNestedResourceMoveExpression[expression] = struct{}{}
 	}
 }
+
+func (checker *Checker) recordReferenceCreation(target, expr ast.Expression) {
+	switch target := target.(type) {
+	case *ast.IdentifierExpression:
+		checker.recordReference(target.Identifier.Identifier, expr)
+	default:
+		// Currently it's not possible to track the references
+		// assigned to member-expressions/index-expressions.
+		return
+	}
+}
+
+func (checker *Checker) recordReference(targetVarName string, expr ast.Expression) {
+	referencedVar := checker.referencedVariable(expr)
+
+	if referencedVar != nil &&
+		referencedVar.Type.IsResourceType() {
+		checker.references.Set(targetVarName, referencedVar)
+	}
+}
+
+// referencedVariable return the referenced variable
+func (checker *Checker) referencedVariable(expr ast.Expression) *Variable {
+	refExpression := checker.referenceExpression(expr)
+
+	var variableRefExpr *ast.Identifier
+
+	switch refExpression := refExpression.(type) {
+	case *ast.ReferenceExpression:
+		// If it is a reference expression, then find the root variable.
+		// i.e: root variable of `&a.b.c as &T` is `a`.
+		// This is because nested resources cannot be moved.
+		// If `c` needs to be moved, then `a` has to be moved.
+		// So tracking `a` is sufficient.
```

Comment:
If field `c` is resource-typed, it can be moved if the field is variable, e.g. through a swap.
As nested resources (here, in a nested field, or e.g. in an array) cannot be tracked, it still makes sense to at least track `a`.

Maybe

```suggestion
		// If it is a reference expression, then find the "root variable".
		// As nested resources cannot be tracked, at least track the "root" if possible.
		// For example, for an expression `&a.b.c as &T`, the "root variable" is `a`.
```